### DISCUSSION TITLE
rust (fd, bat, etc ..)  and java packages are broken on my machine with WSL1

### DISCUSSION THREAD

**Discussion by zodman** [Homebrew/discussions#2777] [Linux] [answered]
for a long time my rust apps like bat, fd and java are broken:

`Linux aUTOMATION 4.4.0-19041-Microsoft #1237-Microsoft Sat Sep 11 14:32:00 PST 2021 x86_64 x86_64 x86_64 GNU/Linux`

```
$ fd
-bash: /home/linuxbrew/.linuxbrew/bin/fd: cannot execute binary file: Exec format error

```

Ok let me check if they are wrongly compiled:

```
 file /home/linuxbrew/.linuxbrew/bin/fd
/home/linuxbrew/.linuxbrew/bin/fd: symbolic link to ../Cellar/fd/8.3.0/bin/fd
file /home/linuxbrew/.linuxbrew/Cellar/fd/8.3.0/bin/fd
/home/linuxbrew/.linuxbrew/Cellar/fd/8.3.0/bin/fd: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /home/linuxbrew/.linuxbrew/lib/ld.so, for GNU/Linux 2.6.32, BuildID[sha1]=d2865f531070f7980744e69a3f7d0c282906a39c, with debug_info, not stripped
```
But If pass:

`/home/linuxbrew/.linuxbrew/lib/ld.so /home/linuxbrew/.linuxbrew/Cellar/fd/8.3.0/bin/fd`

the command works and others too ....

what I miss ?

If I install a package outside brew like bat (download from the release at github).

```
 file   /usr/bin/bat
/usr/bin/bat: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, BuildID[sha1]=87e34e0bb0278eb06cae38b32cc7032fd8a6b36b, stripped
```

the ld.so its different .. I think i missing something from the /etc/ld.so.conf.d/ but I'm not sure ...

```

 ldd    /usr/bin/bat
        linux-vdso.so.1 (0x00007fffd9cfa000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f658c540000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f658c320000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f658c110000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f658bef0000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f658bb50000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f658b940000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f658b530000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f658d000000)
 zodman  ~  ldd  /home/linuxbrew/.linuxbrew/Cellar/fd/8.3.0/bin/fd
        linux-vdso.so.1 (0x00007ffff2196000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f50d1d10000)
        libgcc_s.so.1 => /home/linuxbrew/.linuxbrew/lib/libgcc_s.so.1 (0x00007f50d1cf6000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f50d1ae0000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f50d1740000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f50d1530000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f50d1130000)
        /home/linuxbrew/.linuxbrew/lib/ld.so => /lib64/ld-linux-x86-64.so.2 (0x00007f50d2800000)
 zodman  ~  ls -lah /home/linuxbrew/.linuxbrew/lib/ld.so
lrwxrwxrwx 1 zodman zodman 27 Mar 13  2020 /home/linuxbrew/.linuxbrew/lib/ld.so -> /lib64/ld-linux-x86-64.so.2
 zodman  ~  ls -lah /lib64/ld-linux-x86-64.so.2
lrwxrwxrwx 1 root root 32 Dec  7  2020 /lib64/ld-linux-x86-64.so.2 -> /lib/x86_64-linux-gnu/ld-2.27.so
```

**Comment by carlocab** [comment]
CC @Homebrew/linux

**Comment by danielnachun** [comment, marked as answer]
It appears you are using Windows Subsystem for Linux (WSL).  If you are using WSL v1, this is a well known issue https://github.com/Homebrew/linuxbrew-core/issues/20393, https://github.com/Homebrew/linuxbrew-core/issues/21293.  I've reproduced this myself before, and noticed that someone found this interesting workaround https://github.com/microsoft/WSL/issues/5875#issuecomment-688881367.  Through some digging I did here: https://github.com/Homebrew/linuxbrew-core/pull/22620#issuecomment-854383299, it ultimately may be that this problem is due to a rather arcane issue with some ELF sections using inconsistent byte lengths for alignment: https://github.com/NixOS/patchelf/pull/275.  I suspect that the way WSLv1 emulates dynamic linking can't handle these unusual section alignments (similar to the problem `patchelf` was having), and leads to this error.

There are probably 2 solutions to this issue:
1) Switch to WSLv2 if your hardware supports it.  In some of this issues I linked above, this was confirmed to fix this issue.  WSLv2 provides full Linux syscall and kernel support: https://docs.microsoft.com/en-us/windows/wsl/compare-versions, so it makes sense that things work better there.
2) Rebuild the broken formulae from source.  I'm not sure if this fixes the issue, but if you can't use WSLv2 for some reason it might be worth a try.

**Reply by zodman** [reply]
thanks ... for your reply!
